(django) Re-adds the ChangeLog component

joshwingreene opened the pull request:
## Changes
1. Adds the `ChangeLog` component and adds it to both the `UpdateAgentView` and `UpdateUserView`. Only admins are able to see the ChangeLog.
2. Clicking on the "Show All Changes" button will cause a modal to show up. The user can use this in order to see all of the changes to the user or agent. The newly added `useInfiniteLoading` hook supports this functionality.
3. Created a `ChangeListGroup` component. This is used by both the Changelog component and the Changelog modal.
4. Adds the associated agent and user queries
5. Created a HistoricRecord generic type for the change items
6. Adds the `CircularContainer` styled component
7. Adds relevant theming

## Purpose
Re-adds the Changelog component
- this implementation supports our django backend

## Approach
- In the event that the historyUser is null, I set the UI so that the editorName shows up as 'System'. I checked the value of historyUser when the initial admin is created, and I found that the historyUser is null as well. So, I'm assuming that historyUser is null whenever an agent/user is created or changed without the use of the frontend. 

Here's a screenshot:

![first change object](https://user-images.githubusercontent.com/2876874/197033987-41ade0ba-4cd1-498f-9845-c7ba700d675c.png)

## Additional Notes
- The handleShowAllChanges methods are supposed to include the commented out console log. This is consistent with the previous Changelog PRs.

## Testing Steps
1. Pull in the changes to your local copy of this branch and run it alongside the dj_starter_demo repo.
2. Make some changes to the admin user and then open it in order to see its changelog
3. Create an agent and then make some changes to it. Afterwards, go ahead and open it in order to see its changelog.

## Screenshots

### Admin

![admin - changelog](https://user-images.githubusercontent.com/2876874/197034736-2fc0776c-e578-4692-97a5-01fe6e0ff230.png)

### Agent

![agent - changelog](https://user-images.githubusercontent.com/2876874/197034781-223694e3-ca06-4516-b5a3-599b53a25af6.png)

joshwingreene commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Changelog (connect frontend with backend)](https://trello.com/c/JPoop1MM/101-changelog-connect-frontend-with-backend)

DropsOfSerenity commented on src/common/components/ChangeLog.tsx:
```diff
@@ -0,0 +1,160 @@
+};
+
+const defaultGetChangeItemAction = (action: string): string => {
+  switch (action) {
+    case '+':
+      return 'create';
+    case '~':
+      return 'update';
+    case '-':
+      return 'delete';
+    default:
+      throw Error("Given action does not equal '+', '~', or '-'");
```
We are mapping these to `create`, `update` and `delete` only to transform them later into `created`, `updated` and `deleted` later. Maybe it's best here to just used the past tense names, only since they are the ones that are displayed.

DropsOfSerenity commented on src/common/components/ChangeLog.tsx:
```diff
@@ -0,0 +1,160 @@
+    return `${nameArray[0]} ${nameArray[1].charAt(0)}.`;
+  };
+
+  const getDateStr = (date: Date): string => {
+    return date.toLocaleString(undefined, { month: 'short', day: 'numeric' });
+  };
+
+  const getTimeStr = (date: Date): string => {
+    return date.toLocaleString(undefined, { hour: 'numeric', minute: 'numeric' });
+  };
```
Lets use moment for this.

DropsOfSerenity commented on src/features/agent-dashboard/pages/UpdateAgentView.tsx:
```diff
@@ -10,24 +10,38 @@ import { PageCrumb, PageHeader, SmallContainer } from 'common/styles/page';
+  }, [error, navigate, agentHistoryError]);
+
+  const handleShowAllChanges = () => {
+    // console.log('handleShowAllChanges');
```
Lets display a modal if they want to see all changes, that displays either a paginated or infinite scroll list of changes.

DropsOfSerenity commented on src/common/components/ChangeLog.tsx:
```diff
@@ -0,0 +1,160 @@
+  color: ${props => props.theme.changelogs.accentTextColor};
+`;
+
+const defaultPropertyAccessor = (
+  change: unknown,
+  propertyName: string,
+): string | number | Record<string, unknown> | boolean => {
+  if (isObject(change)) {
+    const value = change[propertyName];
+
+    if (isStringOrNumber(value)) {
+      return value;
+    }
+
+    if (isObject(value)) {
+      return value;
+    }
+
+    if (value === null) {
+      return false;
+    }
+
+    throw Error(`The resulting value from change[${propertyName}] is not a string, number, object, or null`);
+  }
+
+  throw Error('Given change parameter is not an object');
+};
```
I know we had discussed the library version of this before. But I think we should just type this normally, as this adds a layer of indirection that we can add in the future, but isn't necessary up front.

I suggest something like a `HistoricalRecord<T>` generic type.

DropsOfSerenity commented on src/common/components/ChangeLog.tsx:
```diff
@@ -0,0 +1,160 @@
+          <h5 className='mb-3'>Latest Changes</h5>
+          <ListGroup as='ol'>
+            {changeList.map(changeItem => {
+              const id = changePropertyAccessor(changeItem, 'historyId');
+              const changeAction = getChangeItemAction(changePropertyAccessor(changeItem, 'historyType').toString());
+              const creationDate = changePropertyAccessor(changeItem, 'historyDate').toString();
+              const historyUser = changePropertyAccessor(changeItem, 'historyUser');
+
+              let editorName = '';
+
+              if (
+                isObject(historyUser) &&
+                isKeyOfObject('firstName', historyUser) &&
+                isKeyOfObject('lastName', historyUser)
+              ) {
+                editorName = `${historyUser.firstName} ${historyUser.lastName}`;
+              }
```
Ditto on this stuff, although this is a good idea for if/when we turn this into a library. It feels a bit [YAGNI](https://www.martinfowler.com/bliki/Yagni.html) at the moment.

DropsOfSerenity reviewed (changes requested):
This works and seemed easy enough to get going. Would like to simplify this however before we get it in.

joshwingreene commented:
Hey @DropsOfSerenity, This PR is ready for another review.

## Recent Changes
- Resolved your earlier comments
- Added the `useInfiniteLoading` hook in order to make it easy to add the interaction that's used for the Notifications feature. I'm currently using it for the ChangeLog modals.
- Made the history endpoint requests take in a url instead of the id, page, and pageSize
- Moved the rendering of the Changelog's list group to a separate component (`ChangeListGroup`) so that it could be used by both the ChangeLog component and the ChangeLog modal

## Note
I noticed that the useModal hook will cause modals to be rendered outside of the app's context. If you check the inspector, the modal won't be inside of the root element. 

Here's a screenshot:

![modal - outside of app's context](https://user-images.githubusercontent.com/2876874/200962766-e7c8b067-7eab-4163-b7bd-10985ba3ee7a.png)

Because of this, the modal's elements won't get access to the app's context. So, that's why I had to send in the ThemeContext to the `ChangeListGroup` component when that component is being rendered by the ChangeLog modal.

## Demo

https://user-images.githubusercontent.com/2876874/200961197-405e7e34-bee1-4739-89b0-8f38a9a02409.mov

DropsOfSerenity reviewed (changes requested):
> Resolved your earlier comments

Awesome nice work.

> I noticed that the useModal hook will cause modals to be rendered outside of the app's context. If you check the inspector, the modal won't be inside of the root element. Because of this, the modal's elements won't get access to the app's context. So, that's why I had to send in the ThemeContext to the ChangeListGroup component when that component is being rendered by the ChangeLog modal.

Instead just move the ModalProvider to inside the ThemeProvider, so that it has access. Like so.

Before:
![image](https://user-images.githubusercontent.com/2287977/201449747-3a3bbcdc-a924-47c0-8e3f-ec78dd5bfd22.png)

After:
![image](https://user-images.githubusercontent.com/2287977/201449719-a4cf30ab-4792-4868-839c-1f0724fd6f6d.png)

This gets rid of all the extra work you are doing here, as passing the theme context is unnecessary.

This is one of those cases where you've correctly identified a problem (Modal was outside the theme provider), but then implemented a solution that works around the problem rather than fixes it. Don't be afraid to take a wider look at the context of a problem and ask if there is an easier way to do something.

Normally some indicators that you are diving down a bad path, is if you find yourself having to duplicate a lot of knowledge, if an inner components has to know a lot about the working of outer components that's an indication that something is wrong.

I caught this today, I'll give the rest of the branch a more comprehensive review on monday.

DropsOfSerenity commented on src/common/components/ChangeLog/ChangeListGroup.tsx:
```diff
@@ -0,0 +1,82 @@
+        return { changeName: 'Updated', icon: 'pen' };
+      case '-':
+        return { changeName: 'Deleted', icon: 'trash-can' };
+      default:
+        throw Error("Given changeType does not equal '+', '~', or '-'");
```
If we fix the type, this default case is no longer necessary.

DropsOfSerenity commented on src/common/hooks/useInfiniteLoading.ts:
```diff
@@ -0,0 +1,40 @@
+import { PaginatedResult } from 'common/models';
+import { useMemo, useState } from 'react';
+import { UseQuery, UseQueryOptions } from 'rtk-query-config';
+
+export const useInfiniteLoading = <T, ResultType extends PaginatedResult<T>>(
+  initialUrl: string,
+  useQuery: UseQuery<ResultType>,
+  options?: UseQueryOptions,
+) => {
+  const [url, setUrl] = useState<string | null>(initialUrl);
+  const [loadedData, setLoadedData] = useState<T[]>([]);
+
+  const { data, error, isLoading, isFetching } = useQuery(url, options);
+
+  useMemo(() => {
+    if (data && !isLoading) {
+      setLoadedData(n => [...n, ...data.results]);
+    }
+  }, [data, isLoading]);
+
+  const hasMore = useMemo(() => {
+    if (isLoading || isFetching) return false;
+    return !!data?.links.next;
+  }, [data, isLoading, isFetching]);
+
+  const fetchMore = () => {
+    if (hasMore && data) {
+      setUrl(data.links.next);
+    }
+  };
+
+  return {
+    loadedData,
+    error,
+    isFetching,
+    totalCount: data?.meta.count,
+    hasMore,
+    fetchMore,
+  };
+};
```
Nice work on this, this seems nicely usable :)

DropsOfSerenity commented on src/features/user-dashboard/pages/UpdateUserView.tsx:
```diff
@@ -1,39 +1,98 @@
   const availableRoles = roles.filter(role => userHasPermission({ permission: 'role:read', data: role }));
   const [formValidationErrors, setFormValidationErrors] = useState<ServerValidationErrors<FormData> | null>(null);
 
+  const [showModal, hideModal] = useModal(
+    ({ in: open, onExited }) => {
+      return (
+        <Modal show={open} onHide={hideModal} onExited={onExited}>
+          <Modal.Header closeButton>
+            <Modal.Title>Changes</Modal.Title>
+          </Modal.Header>
+          <Modal.Body>
+            <ChangeListGroup changeList={userHistory} />
+          </Modal.Body>
+          {hasMore && (
+            <Modal.Footer className='d-flex justify-content-center'>
+              <LoadingButton
+                className='action-shadow'
+                loading={isFetchingHistory}
+                variant='primary'
+                onClick={() => fetchMore()}
+              >
+                Load More
+              </LoadingButton>
+            </Modal.Footer>
+          )}
+        </Modal>
+      );
+    },
+    [userHistory, isFetchingHistory],
+  );
```
Lets give this modal a max height and make it so that it can scroll inside the modal itself rather than the whole page scrolling.

![image](https://user-images.githubusercontent.com/2287977/202551923-55977e21-0889-4e11-bc26-a6e5d7b4b625.png)

DropsOfSerenity reviewed (changes requested):
Overall this looks great! Nice work, I have a few minor tweaks that we need before we can get this in, but nice work!

joshwingreene commented:
@DropsOfSerenity I attached a video of the updated experience for the changelog modal. I was also able to fix the issue that was causing the browser's scollbar to be applied to all of the modals. For that issue, I removed the min-height setting from the modal-dialog class and used the React-Bootstrap's centered prop.

https://user-images.githubusercontent.com/2876874/203444403-0e684a9d-34bf-41e9-af8e-c1bbc2b6aee8.mov

DropsOfSerenity reviewed (approved):
Alright nice work! I think I'm good with this getting in now @joshwingreene
